### PR TITLE
`metrics.go` support for metadata queries(labels and series)

### PR DESCRIPTION
**Pull request by kavirajk**
**What this PR does / why we need it**:
We use `metrics.go`  (a convention) to record some useful statistics about each of the `instance` and `range` queries. 

So that we can query something like
`{foo="bar"}|= "metrics.go" |logfmt|latency=slow `

Where `latency` comes from statistics recorded in `pkg/logql/metric.go`.

This PR add support those experience for series and labels API as well.

Examples:
1. Labels API (names)
```
curl -G -s  "http://127.0.0.1:3100/loki/api/v1/label"
```

Generates following log lines via `metrics.go`
```
level=info ts=2022-05-03T19:51:37.417252626Z caller=metrics.go:170 component=frontend org_id=145265 traceID=5bdfb2e229890ea2 latency=fast query_type=labels length=1h0m0s duration=65.140762ms status=200 label= throughput=0B total_bytes=0B total_entries=14
```

2. Labels API (values)
```
curl -G -s  "http://127.0.0.1:3100/loki/api/v1/label/container/values"
```

Generates following log lines via `metrics.go`
```
level=info ts=2022-05-03T19:50:07.710590524Z caller=metrics.go:170 component=frontend org_id=145265 traceID=11349be1207cbdba latency=fast query_type=labels length=1h0m0s duration=11.239903ms status=200 label=container throughput=0B total_bytes=0B total_entries=19
```

3. Series API
```
curl -G -s  "http://127.0.0.1:3100/loki/api/v1/series" --data-urlencode 'match[]={app="loki"}'
```

Generates following log lines via `metrics.go`
```
level=info ts=2022-05-03T19:51:40.272480868Z caller=metrics.go:215 component=frontend org_id=145265 traceID=396c8cdce34bb53e latency=fast query_type=series length=1h0m0s duration=85.968193ms status=200 match="{app=\"loki\"}" throughput=0B total_bytes=0B total_entries=249
```

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/loki/issues/4503

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

**Future Work**
- [ ] Get rid of `latency=slow|fast` label from metric.go entries.
- [ ] deprecate and remove `returned_lines` when `query_type=filtered`. Because now we have `total_entries` common for all the `query_type`.

### CODE REVIEW

File: pkg/logql/metrics.go [hunk 1]
```diff
@@ -13,6 +14,7 @@ import (
 
 	"github.com/grafana/loki/pkg/logql/syntax"
 	"github.com/grafana/loki/pkg/logqlmodel"
+	"github.com/grafana/loki/pkg/logqlmodel/stats"
```
Comment (cstyan) [inline pkg/logql/metrics.go:17]: extra import, line below is `logql_stats "github.com/grafana/loki/pkg/logqlmodel/stats"`

File: pkg/logql/metrics.go [hunk 2]
```diff
@@ -138,6 +142,95 @@ func RecordMetrics(ctx context.Context, log log.Logger, p Params, status string,
...
+	execLatency.WithLabelValues(status, queryType, "").
+		Observe(stats.Summary.ExecTime)
+	chunkDownloadLatency.WithLabelValues(status, queryType, "").
+		Observe(stats.ChunksDownloadTime().Seconds())
+	duplicatesTotal.Add(float64(stats.TotalDuplicates()))
+	chunkDownloadedTotal.WithLabelValues(status, queryType, "").
+		Add(float64(stats.TotalChunksDownloaded()))
+	ingesterLineTotal.Add(float64(stats.Ingester.TotalLinesSent))
+
+	// TODO(kavi): Do we need recordUsageStats for labels query?
```
Comment (cstyan) [inline pkg/logql/metrics.go:186]: these are new to me, we use expvar? :thinking:

File: pkg/logql/metrics.go [hunk 3]
```diff
@@ -138,6 +142,95 @@ func RecordMetrics(ctx context.Context, log log.Logger, p Params, status string,
...
+		queryType   = QueryTypeLabels
+	)
+
+	// Tag throughput metric by latency type based on a threshold.
+	// Latency below the threshold is fast, above is slow.
+	if stats.Summary.ExecTime > slowQueryThresholdSecond {
+		latencyType = latencyTypeSlow
+	}
+
+	// we also log queries, useful for troubleshooting slow queries.
```
Comment (cstyan) [inline pkg/logql/metrics.go:163]: I think this comment isn't necessary?

File: pkg/logql/metrics.go [hunk 4]
```diff
@@ -138,6 +142,95 @@ func RecordMetrics(ctx context.Context, log log.Logger, p Params, status string,
...
+		queryType   = QueryTypeSeries
+	)
+
+	// Tag throughput metric by latency type based on a threshold.
+	// Latency below the threshold is fast, above is slow.
+	if stats.Summary.ExecTime > slowQueryThresholdSecond {
+		latencyType = latencyTypeSlow
+	}
+
+	// we also log queries, useful for troubleshooting slow queries.
```
Comment (cstyan) [inline pkg/logql/metrics.go:208]: same here

**Review by cstyan — COMMENTED** [review]
Would we want any additional stats fields for these queries? the # of results for example

File: pkg/logql/metrics.go (hunk 2 above)
Comment (kavirajk) [reply pkg/logql/metrics.go:183]: yes. we expvar for the usage stats.

**Comment by kavirajk** [comment]
> Would we want any additional stats fields for these queries? the # of results for example

Good point. I think that would really be helpful. Let me try adding it.

File: pkg/logql/metrics.go (hunk 2 above)
Comment (cstyan) [reply pkg/logql/metrics.go:183]: with what though? how do we collect those? what's the reason they aren't just an internal go struct (if we collect them just from within our own go processes) or prometheus metrics if we need to see the values across processes?

Comment (kavirajk) [reply pkg/logql/metrics.go]: I'm not familiar with [usageStats](https://github.com/grafana/loki/tree/main/pkg/usagestats) either :). 

Removing the TODO, as it's not the scope of this PR to add usage stats to series and labels API. We can add it later in separate PR.

**Review by trevorwhitney — COMMENTED** [review]
Looks great!

**Review by cstyan — APPROVED** [review]
LGTM as well, there's just a lint issue it looks like.

Thanks for implementing this Kavi. I learned a thing or two about how the stats collection is implemented while reviewing.

File: pkg/querier/queryrange/codec.go [hunk 5]
```diff
@@ -779,8 +779,16 @@ func paramsFromRequest(req queryrangebase.Request) (logql.Params, error) {
...
+		return &paramsSeriesWrapper{
+			LokiSeriesRequest: r,
+		}, nil
+	case *LokiLabelNamesRequest:
+		return &paramsLabelNamesWrapper{
+			LokiLabelNamesRequest: r,
+		}, nil
 	default:
-		return nil, fmt.Errorf("expected *LokiRequest or *LokiInstantRequest, got (%T)", r)
+		return nil, fmt.Errorf("expected one of the *LokiRequestor, *LokiInstantRequest, *LokiSeriesRequest, *LokiLabelNamesRequest, got (%T)", r)
```
Comment (owen-d) [inline pkg/querier/queryrange/codec.go:791]:
```suggestion
		return nil, fmt.Errorf("expected one of *LokiRequest, *LokiInstantRequest, *LokiSeriesRequest, *LokiLabelNamesRequest, got (%T)", r)
```

**Review by owen-d — COMMENTED** [review]
The approach is looking good. As you mentioned, we'll need to find/fix the panic and do some linting.

**Comment by kavirajk** [comment]
This is how the metrics.go entries looks like for following metadata queries.(tested it on some internal loki clusters)
1. series
2. label names
3. label values (notice a value for `label` in metrics.go)

Screenshot_2022-05-03_21-55-25

File: pkg/querier/queryrange/stats.go [hunk 6]
```diff
@@ -91,39 +116,85 @@ func StatsCollectorMiddleware() queryrangebase.Middleware {
...
 
+func getLabelNameFromLabelsQuery(path string) string {
+	if strings.HasSuffix(path, "/values") {
+
+		toks := strings.FieldsFunc(path, func(r rune) bool {
+			return r == '/'
+		})
+
+		// now assuming path has suffix `/values` label name should be second last to the suffix
+		// **if** there exists the second last.
```
Comment (cstyan) [inline pkg/querier/queryrange/stats.go:187]: can we include a link to the docs for this endpoint? https://grafana.com/docs/loki/latest/api/#get-lokiapiv1labelnamevalues

**Review by owen-d — APPROVED** [review]
LGTM

**Pull request merged**